components/browser/menu/src/main/java/mozilla/components/browser/menu/BrowserMenu.kt

```diff
@@ -172,114 +229,48 @@ open class BrowserMenu internal constructor(
 }
 
 @VisibleForTesting
-internal fun PopupWindow.displayPopup(
-    containerView: View,
-    anchor: View,
-    preferredOrientation: BrowserMenu.Orientation
-) {
-    // Measure menu
-    val spec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
-    containerView.measure(spec, spec)
-
-    val (availableHeightToTop, availableHeightToBottom) = getMaxAvailableHeightToTopAndBottom(anchor)
-    val containerHeight = containerView.measuredHeight
-
-    val fitsUp = availableHeightToTop >= containerHeight
-    val fitsDown = availableHeightToBottom >= containerHeight
-
+internal fun PopupWindow.displayPopup(currentData: MenuPositioningData) {
     // Try to use the preferred orientation, if doesn't fit fallback to the best fit.
-    when {
-        preferredOrientation == DOWN && fitsDown -> {
-            showPopupWithDownOrientation(anchor)
-        }
-        preferredOrientation == UP && fitsUp -> {
-            showPopupWithUpOrientation(anchor, availableHeightToBottom, containerHeight)
-        }
-        else -> {
-            showPopupWhereBestFits(
-                anchor,
-                fitsUp,
-                fitsDown,
-                availableHeightToTop,
-                availableHeightToBottom,
-                containerHeight
-            )
-        }
-    }
-}
+    when (currentData.inferredMenuPlacement) {
+        is BrowserMenuPlacement.AnchoredToTop.Dropdown -> showPopupWithDownOrientation(currentData)
+        is BrowserMenuPlacement.AnchoredToBottom.Dropdown -> showPopupWithUpOrientation(currentData)
 
-@Suppress("LongParameterList")
-private fun PopupWindow.showPopupWhereBestFits(
-    anchor: View,
-    fitsUp: Boolean,
-    fitsDown: Boolean,
-    availableHeightToTop: Int,
-    availableHeightToBottom: Int,
-    containerHeight: Int
-) {
-    // We don't have enough space to show the menu UP neither DOWN.
-    // Let's just show the popup at the location of the anchor.
-    if (!fitsUp && !fitsDown) {
-        showAtAnchorLocation(anchor, availableHeightToTop < availableHeightToBottom)
-    } else {
-        if (fitsDown) {
-            showPopupWithDownOrientation(anchor)
-        } else {
-            showPopupWithUpOrientation(anchor, availableHeightToBottom, containerHeight)
-        }
+        is BrowserMenuPlacement.AnchoredToTop.ManualAnchoring,
+        is BrowserMenuPlacement.AnchoredToBottom.ManualAnchoring -> showAtAnchorLocation(currentData)
     }
 }
 
-private fun PopupWindow.showPopupWithUpOrientation(anchor: View, availableHeightToBottom: Int, containerHeight: Int) {
+@VisibleForTesting
+internal fun PopupWindow.showPopupWithUpOrientation(menuPositioningData: MenuPositioningData) {
+    val anchor = menuPositioningData.inferredMenuPlacement!!.anchor
```

Comment:
Let's try to avoid '!!' as it could cause Null Pointer Exception, when the value is not available lets use a default value.

Reply:
Agree that it doesn't look nice but the NPE really shouldn't happen.
The way I see it, that property is nullable just because I had to use 'MenuPositioningData' data class before actually having a value for the placement property.
But by using an `if-else` [here](https://github.com/mozilla-mobile/android-components/pull/9879/files/e44e91d70001829bcee737787415d78ecc10bb19#diff-464735d6ea95e8651da7c30537532310b3335851468409b0d238eb6ed6fcae16R70-R88) it should always have a valid value. Except when that method to infer it is not called, something that we'd want to catch and fix as soon as possible so crashing in that case may be desired offering early in development the chance to fix it.
Thoughts?

Reply:
I completely understand you point, in my opinion I would rather prefer not using '!!' as it will required a pre-validation to happen, if the code is updated without taking the pre-validation into account, we could have a potential crash. Having a default value and updating it when it's needed made the code safe by default.